### ISSUE TITLE
Box "builder"

### ISSUE THREAD

**Issue by Obvio** [wplib/wplib-box#120]
Add UI for pre selecting certain configuration options, like domain, directory, initial plugins/themes to install etc.

Add the options to "save" a configuration as my default config so next time I'll start with the same base site.

An example of builder/customiser can be seen here http://www.initializr.com/ (click the **responsive** button).

**Comment by mikeschinkel** [comment]
Hi Amit,

> Add UI for pre selecting certain configuration options, like domain

Local domain and/or domain to be used when the site goes live?

> directory

We had planned to standardize on a directory structure for things we'd generate, below; how would you want to do it differently?  Would doing it differently be important?

**On Developer's (Host's) Computer**

- `{project_root}/`
- `{project_root}/www`  <-- WordPress goes here
- `{project_root}/sql`     <-- Databases
- `{project_root}/scripts`  <-- Scripts for the box
- `{project_root}/tests`  <-- Test scripts
- `{project_root}/docs`  <-- Documentation

**In WPLib Box (Guest Computer)**

- `/var/www/`
- `/var/www/www`  <-- WordPress goes here
- `/var/www/sql`     <-- Databases
- `/var/www/scripts`  <-- Scripts for the box
- `/var/www/tests`  <-- Test scripts
- `/var/www/docs`  <-- Documentation

We may actually move from `/var/www` to `/www`.  If we can standardize on these it will allow us to get more things done more quickly and with less effort.

More questions to come...






,, initial plugins/themes to install etc., Ideally the user would have something similar to the wp-admin/themes and wp-admin/plugins type of search where he can pull those in.

**Comment by Obvio** [comment]
in 'directory' i had the assumption many sites could live under one installation, and by 'domain' i ment local developemnt domain, so given this diretcory structure i would say domain1.box maps to www/domain1-wp-files.

that would allow multiple sites and easy 'differentiation' between sites when deploying for example.

**Comment by mikeschinkel** [comment]
@Obvio 

Ah, I was juggling several things when I posted early so I forgot to mention that we plan both 1.) single-site mode and 2.) multi-site mode.  Above is for 1, single-site mode.

For multi-site we plan _(per site):_

**On Developer's (Host's) Computer**

- `{project_root}/sites/{hostname}/`             <-- Project root
- `{project_root}/sites/{hostname}/www`     <-- WordPress goes here
- `{project_root}/sites/{hostname}/sql`        <-- Databases
- `{project_root}/sites/{hostname}/scripts`  <-- Scripts for the box
- `{project_root}/sites/{hostname}/tests`     <-- Test scripts
- `{project_root}/sites/{hostname}/docs`     <-- Documentation

**In WPLib Box (Guest Computer)**

- `/vagrantsites/{hostname}/`              <-- Project root
- `/var/sites/{hostname}/www`   <-- WordPress goes here
- `/vagrant/sites/{hostname}/sql`        <-- Databases
- `/vagrant/sites/{hostname}/scripts`  <-- Scripts for the box
- `/vagrant/sites/{hostname}/tests`     <-- Test scripts
- `/vagrant/sites/{hostname}/docs`     <-- Documentation

Using the API, CLI or Console you chould switch back and forth between Single and Multi-site mode but not sure how we will handle switching back to single site mode it you have multiple sites in multi-site mode; we may just disallow?

**Comment by mikeschinkel** [comment]
> _Initial plugins/themes to install etc., Ideally the user would have something similar to the wp-admin/themes and wp-admin/plugins type of search where he can pull those in._

Trying to understand this requirement.  It seems there are many places to search for plugins and themes, should we really try to duplicate that functionality ourselves in our console?  This seems like it could be an entire website designed around searching for plugins and themes which is a very big projects.  And as I am sure you have seem WordPress.org is currently reinventing their own plugin section.

Our are you thinking something much smaller in scope?

**Comment by mikeschinkel** [comment]
@Obvio:
> _Initial plugins/themes to install etc., Ideally the user would have something similar to the wp-admin/themes and wp-admin/plugins type of search where he can pull those in._

Trying to understand this requirement.  It seems there are many places to search for plugins and themes, should we really try to duplicate that functionality ourselves in our console?  This seems like it could be an entire website designed around searching for plugins and themes which is a very big projects.  And as I am sure you have seem WordPress.org is currently reinventing their own plugin section.

Our are you thinking something much smaller in scope?

**Comment by mikeschinkel** [comment]
@Obvio:

> _Add the options to "save" a configuration as my default config so next time I'll start with the same base site._

So I am clear, you mean you want whatever the current box configuration to be exportable and importable?  How about also automatically saved?

_(BTW, I think we have several different feature requests here and it will be easier for us to track them as separate tickets so we may break them out once we understand them.)_

**Comment by Obvio** [comment]
you're right about not trying to duplicate the .org or themes/plugins search functionality from within wp-admin. 

Than this feature could be based on creating an empty site with my custom configuration, installing the themes and plugins I need and then just use this as the "clone base".

As for switching between single and multisite, not sure about it at all.. on my use cases I just need to have separate wp installations each on its own dev-domain.

**Comment by mikeschinkel** [comment]
> Than this feature could be based on creating an empty site with my custom configuration, installing the themes and plugins I need and then just use this as the "clone base".

Would a list of plugin slugs and a theme slug from WordPress.org, or an HTTP(S) URL work for this?

> As for switching between single and multisite, not sure about it at all.. on my use cases I just need to have separate wp installations each on its own dev-domain.

Many sites in one WPLib Box, right?  

_(You can currently do this with multi-tenancy, but it is manual. Let me know if you want to know how.  Message me on FB your email address if you want to discuss on our support chat on Slack, or you can [discuss on our Facebook group](https://www.facebook.com/groups/wplib/).)_

**Comment by Obvio** [comment]
Yes - a list of plugin slugs and a theme slug from WordPress.org, or an HTTP(S) URL work for this great.

Yes for the second one as well, multi tenancy is what I'm looking for, thanks for the offer, i'll reach out on slack :)
